DOCS-5144 Adds Build an Integration, Build a Marketplace Offering, and Create a Tile Documentation (#14297)
DataDog/integrations-core: docs/dev/agent_integration.md

  - /developers/integrations/testing/
  - /integrations/datadog_checks_dev/
  - /guides/new_integration/
  - /developers/integrations/new_check_howto/
dependencies: "https://github.com/DataDog/integrations-core/blob/alai97/add-marketplace-documentation/docs/dev/agent_integration.md"
---

## Overview

> This guide provides instructions for creating a Datadog Agent integration. For more information about why you would want to create an Agent-based integration, see [Creating your own solution][1].

Reviewer: Sending a user off to the middle section of a new page in the second sentence may cause some whiplash, where the user feels like they are bouncing around. 

Also, it's not clear to me what "Creating your own solution" means or why it matters. 

In other words, how does a user know if they should keep reading, or follow the link instead?

Relatedly, why wouldn't we just tell the user on this page why they would want to create an Agent-based integration?

Reviewer 2: @anawishnoff since we already mention this on the landing page, should we remove this line altogether?

Reviewer 3: Yeah, let's remove this line.

  - /developers/integrations/testing/
  - /integrations/datadog_checks_dev/
  - /guides/new_integration/
  - /developers/integrations/new_check_howto/
dependencies: "https://github.com/DataDog/integrations-core/blob/alai97/add-marketplace-documentation/docs/dev/agent_integration.md"
---

## Overview

This guide provides instructions for creating a Datadog Agent integration. You can list an Agent integration for a price on the Datadog Marketplace, or list it out-of-the-box on the Integrations page. 

## Development Process

You can expect the following process for building an Agent-based integration:

> 1. Meet with the Datadog Technology Partner team to discuss your offering and use cases.

Reviewer 4: ```suggestion
1. Once you've been accepted to the Datadog partner network, you'll meet with the Datadog Technology Partner team to discuss your offering and use cases.
```

2. Request a sandbox account for development.
3. Begin development of your integration, which will include writing the integration code on your end as well as building and installing a python wheel (.whl).
4. Test your integration in your sandbox account.

> 5. Once your development work is tested and complete, follow the steps to [Create a Tile][24] in order to display your offering on the Marketplace or Integrations page.

Reviewer 4: Not sure if we should include a link to the 'Create a Tile' page since the overview section of that page specifically calls out 'For all offerings that do not use the Datadog Agent,....'

  - /developers/integrations/testing/
  - /integrations/datadog_checks_dev/
  - /guides/new_integration/
  - /developers/integrations/new_check_howto/

> dependencies: "https://github.com/DataDog/integrations-core/blob/alai97/add-marketplace-documentation/docs/dev/agent_integration.md"

Author: ```suggestion
dependencies: "https://github.com/DataDog/integrations-core/blob/master/docs/dev/agent_integration.md"
```

further_reading:

> - link: "https://www.datadoghq.com/blog/programmatically-manage-your-datadog-integrations/"

Reviewer: Can we remove this blog from Further Reading? It's 5 years old and doesn't touch on Agent integrations. 

Could we instead add the "Create a Tile" page and whichever other resources we are pointing the user to in the main body of the documentation?

Reviewer 3: @alai97 leaving this one up to you

> - link: "https://www.datadoghq.com/blog/programmatically-manage-your-datadog-integrations/"
>   tag: "Blog"
>   text: "Programmatically manage your Datadog integrations"

Author: ```suggestion
- link: "/developers/integration/create_a_tile/"
  tag: "Documentation"
  text: "Create a tile"
```

- link: "/developers/integrations/python/"
  tag: "Documentation"
  text: "Python for Agent-based Integration Development"
- link: "/developers/"
  tag: "Documentation"
  text: "Learn how to develop on the Datadog platform"
---

## Overview

This page walks Technology Partners through how to create a Datadog Agent integration, which you can list as out-of-the-box on the [Integrations page][23], or for a price on the [Marketplace page][24]. 

## Agent-based integrations

> Agent-based integrations use the [Datadog Agent][17] to submit data through checks written by the developer. Checks can emit [metrics][23], [events][24], and [service checks][25] into a customer's Datadog account. The Agent itself can submit [logs][26] as well, but that is configured outside of the check itself. 

Reviewer: ```suggestion
Agent-based integrations use the [Datadog Agent][17] to submit data through checks written by the developer. Checks can emit [metrics][23], [events][24], and [service checks][25] into a customer's Datadog account. The Agent itself can submit [logs][26] as well, but that is configured outside of the check. 
```

The implementation code for these integrations is hosted by Datadog. Agent integrations are best suited for collecting data from systems or applications that live in a local area network (LAN) or virtual private cloud (VPC). Creating an Agent integration requires you to publish and deploy your solution as a Python wheel (`.whl`).

> You can include out-of-the-box assets such as [monitors][27], [dashboards][28], and [log pipelines][29] with your Agent-based integration. When a user clicks **Install** on your integration tile, they are prompted to follow the setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the their account.

Reviewer 3: Clarified this to remove the bit about recommended monitors, call out that dashboards appear right away after install and pipelines will appear after proper installation/setup

> ## Development process

Reviewer: For any changes we make below, let's be sure mirror them in the corresponding section on the "Create an API" page

The process to build an Agent-based integration looks like the following: 

> 1. Once you've been accepted to the Datadog Partner Network, you will meet with the Datadog Technology Partner team to discuss your offering and use cases.

Reviewer: Can we link to Datadog Partner Network here?

> 2. Request a Datadog sandbox account for development.

Reviewer: ```suggestion
2. Request a Datadog sandbox account for development through your Datadog Partner Network account.
```

3. Begin development of your integration, which includes writing the integration code on your end as well as building and installing a Python wheel (`.whl`).
4. Test your integration in your Datadog sandbox account.

> 5. Once your development work is tested and complete, follow the steps in [Create a Tile][20] to display your offering on the **Integrations** or **Marketplace** page.

Reviewer: It seems like we are still asking partners to run ddev to get the scaffolding to build an Agent integration, but then asking them on the "Create a Tile" page to rerun ddev creating another set of scaffolding. 

The scaffolding the partner creates for their Agent integration has everything they need for their tile. It'll likely be confusing to point them to documentation that asks them to create a directory run ddev again.

> 5. Once your development work is tested and complete, follow the steps in [Create a Tile][20] to display your offering on the **Integrations** or **Marketplace** page.

Reviewer 3: ```suggestion
5. Once your development work is tested and complete, populate your tile assets (including setup instructions, images, and other metadata) in order to display your offering on the **Integrations** or **Marketplace** page.
```

- [Docker][4] to run the full test suite.
- The git [command-line][5] or [GitHub desktop client][19].

PAID

> {{% tab "Paid integrations" %}}

Reviewer 2: ```suggestion
{{% tab "Marketplace integrations" %}}
```

## Set up a directory and clone the Marketplace repository

> Once you've decided on an offering, set up a directory:

Reviewer 2: ```suggestion
Set up a directory:
```

   {{< code-block lang="shell" >}}git clone git@github.com:DataDog/marketplace.git{{< /code-block >}}
4. Create a feature branch to work in.
    git switch -c <YOUR INTEGRATION NAME> origin/master

## Install and configure the Datadog development toolkit

The Agent Integration Developer Tool allows you to create scaffolding when you are developing an integration by generating a skeleton of your integration tile's assets and metadata. For instructions on installing the tool, see [Install the Datadog Agent Integration Developer Tool][25].

After you install the Developer tool, configure it for the `marketplace` repo:

Set `marketplace` as the default working repository:

{{< code-block lang="shell" >}}
ddev config set marketplace $HOME/dd/marketplace
ddev config set repo marketplace
{{< /code-block >}}

If you used a directory other than `$HOME/dd` to clone the marketplace directory, use the following command to set your working repository:

{{< code-block lang="shell" >}}
ddev config set marketplace <PATH/TO/MARKETPLACE>
ddev config set repo marketplace
{{< /code-block >}}

{{% /tab %}}

> {{% tab "Free Integrations" %}

Reviewer 2: ```suggestion
{{% tab "Out-of-the-box Integrations" %}
```

- [Docker][4] to run the full test suite.
- The git [command-line][5] or [GitHub desktop client][19].

> You can list an Agent integration for a price in the Datadog Marketplace, or list it out-of-the-box on the Integrations page.

Reviewer: Is it possible to make this section more detailed and actionable? 

For instance, after we explain the differences between the Marketplace and Integrations page options, can we add an imperative sentence that directs the user to "choose" one of the integration paths below using the buttons?

Reviewer 2: @anawishnoff Moving this to the overview section might make it standout more and we can add actionable items.

Reviewer 3: Yeah I agree with Alex, let's add this in after explaining the differences, and include a call to action (i.e. select an option below to get started)

> {{< tabs >}}

Reviewer 4: love this!

> {{% tab "Marketplace Integrations" %}}

Reviewer: Can we change this to something clearer? Like "Build Marketplace Agent Integration" or just "Build Marketplace Integration" if it needs to be shorter?

Same with the Out-of-the-box button

## Set up a directory and clone the Marketplace repository

Set up a directory:

1. Request access to the [Marketplace repository][101] by following the instructions in the [Marketplace documentation][102].
2. Create a `dd` directory:
   {{< code-block lang="shell" >}}mkdir $HOME/dd{{< /code-block >}}

   The Datadog Development Toolkit command expects you to be working in the `$HOME/dd/` directory. This is not mandatory, but working in a different directory requires additional configuration steps.

> 3. Once you have been granted access to the Marketplace repository, create the `dd` directory and clone the `marketplace` repo:

Reviewer 4: Preview looks funny here. 
![Uploading Screen Shot 2023-04-25 at 8.34.14 AM.png…]()

   {{< code-block lang="shell" >}}git clone git@github.com:DataDog/marketplace.git{{< /code-block >}}
4. Create a feature branch to work in.
    {{< code-block lang="shell" >}}git switch -c <YOUR INTEGRATION NAME> origin/master{{< /code-block >}}

## Install and configure the Datadog development toolkit

The Agent Integration Developer Tool allows you to create scaffolding when you are developing an integration by generating a skeleton of your integration tile's assets and metadata. For instructions on installing the tool, see [Install the Datadog Agent Integration Developer Tool][103].

After you install the Developer tool, configure it for the `marketplace` repo:

> Set `marketplace` as the default working repository:

Reviewer 4: Maybe this can be combined into one line

Reviewer 2: Going to match it to ootb integrations directions. Numbered set and set working repo

Reviewer 2: The reason it isn't one line is because a partner may have set the repo in a directory other than home. @anawishnoff I agree this does look weird. We can just say "After you install the Developer tool, configure it for the `marketplace` repo by following the instructions below"

Reviewer 3: That sounds good

{{< code-block lang="shell" >}}
ddev config set marketplace $HOME/dd/marketplace
ddev config set repo marketplace
{{< /code-block >}}

If you used a directory other than `$HOME/dd` to clone the marketplace directory, use the following command to set your working repository:

{{< code-block lang="shell" >}}
ddev config set marketplace <PATH/TO/MARKETPLACE>
ddev config set repo marketplace
{{< /code-block >}}

[101]: https://github.com/Datadog/marketplace
[102]: https://docs.datadoghq.com/developers/integrations/marketplace_offering
[103]: https://docs.datadoghq.com/developers/integrations/python

{{% /tab %}}

{{% tab "Out-of-the-box Integrations" %}}

> ## Set up your integrations-extra repo

Reviewer: Is "Set up your" accurate here? Would "fork and clone the integrations-extras repository" or something similar be more specific? It's not their repo. They are just making a copy of the existing repo locally.

Reviewer 2: We mention step by step in the instructions to fork and clone, I think including that at the title might lead partners to do that rather than following directions. This also keeps consistency with the marketplace instructions.

Reviewer: I'm just calling out that, semantically, the user is not setting up their own integrations-extra repo, so I find the heading confusing. 

Same with the next line: "Follow these instructions to set up your repo for integration development"

Datadog owns the int-extras repo. We're only asking the user to set up their own _directory_ where they will fork and clone a copy of the int-extras repo

- [Docker][4] to run the full test suite.
- The git [command-line][5] or [GitHub desktop client][19].

**To get started, select either tab below for instructions on building an out-of-the-box Agent-based integration or an Agent-based integration on the Marketplace.** 

> {{< tabs >}}

Reviewer 4: Should we have OOTB integration as the default selection?

{{% tab "Build a Marketplace Integration" %}}
## Set up a directory and clone the Marketplace repository

Set up a directory:

1. Request access to the [Marketplace repository][101] by following the instructions in the [Marketplace documentation][102].
2. Create a `dd` directory:
   {{< code-block lang="shell" >}}mkdir $HOME/dd{{< /code-block >}}

   The Datadog Development Toolkit command expects you to be working in the `$HOME/dd/` directory. This is not mandatory, but working in a different directory requires additional configuration steps.

3. Once you have been granted access to the Marketplace repository, create the `dd` directory and clone the `marketplace` repo:
   {{< code-block lang="shell" >}}git clone git@github.com:DataDog/marketplace.git{{< /code-block >}}

> 4. Create a feature branch to work in.

Reviewer: This font looks smaller than for other steps. It also doesn't have a space after it before the next heading. 
![image](https://user-images.githubusercontent.com/116011073/235536280-65a08976-6cfd-4d5e-be89-e1948530db4d.png)

   ddev config set repo marketplace
   {{< /code-block >}}

[101]: https://github.com/Datadog/marketplace
[102]: https://docs.datadoghq.com/developers/integrations/marketplace_offering
[103]: https://docs.datadoghq.com/developers/integrations/python

{{% /tab %}}

{{< /tabs >}}

## Create your integration

Once you've downloaded Docker, installed an appropriate version of Python, and prepared your development environment, you can start creating an Agent-based integration. 

The following instructions use an example integration called `Awesome`. Follow along using the code from Awesome, or replace Awesome with your own code, as well as the name of your integration within the commands. For example, use `ddev create <your-integration-name>` instead of `ddev create Awesome`. 

### Create scaffolding for your integration

The `ddev create` command runs an interactive tool that creates the basic file and path structure (or scaffolding) necessary for an Agent-based integration.

1. Before you create your first integration directory, try a dry-run using the `-n/--dry-run` flag, which doesn't write anything to the disk:
   ```
   ddev create -n Awesome
   ```

>    This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` or `marketplace` repository location.

Reviewer 3: I'm fine with just removing the repo names completely

   ddev create -n Awesome
   ```

>    This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` or `marketplace` repository location.

Reviewer: This happens again later. For example, even though the selected "Marketplace Integrations," under "Write validation checks" it says: "Tests are required if you want your integration to be included in the **integrations-extras** repository."

>    This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` or `marketplace` repository location.

Reviewer 3: Hey @alexvetras, the tabs only apply to the sections:
- Set up a directory and clone the Marketplace repository
- Install and configure the Datadog development toolkit

All of the content after those two sections is shared.

>    This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` or `marketplace` repository location.

Reviewer 2: @alexvetras Do you mean including this section in the tabs and changing the wording based off which tab/path they chose?

   ddev create -n Awesome
   ```

   This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` or `marketplace` repository location.

> 1. Run the command without the `-n` flag. The tool asks you for an email and name and then creates the files you need to get started with an integration.

Reviewer: Do we have any guidance on how partners should name their integrations? For example, in the Marketplace, we need users to follow a {partner}.{integration technology} convention, because more than one partner could create an integration for the same technology. 

Also, if the user runs the command with the integration name "awesome" in the example, but then decides to use that as a template to create their own check, where will they need to change out the name "awesome"? Is it only picked up by the name of a directory, or is it added in other places? I just want to ensure that if all users start by  creating "awesome" checks, that it doesn't causes issues when they go to open a pull-request for their customer integration.

## Create your integration

Once you've downloaded Docker, installed an appropriate version of Python, and prepared your development environment, you can start creating an Agent-based integration. 

The following instructions use an example integration called `Awesome`. Follow along using the code from Awesome, or replace Awesome with your own code, as well as the name of your integration within the commands. For example, use `ddev create <your-integration-name>` instead of `ddev create Awesome`. 

### Create scaffolding for your integration

The `ddev create` command runs an interactive tool that creates the basic file and path structure (or scaffolding) necessary for an Agent-based integration.

1. Before you create your first integration directory, try a dry-run using the `-n/--dry-run` flag, which doesn't write anything to the disk:
   ```
   ddev create -n Awesome
   ```

>    This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `marketplace` repository location.

Author: ```suggestion
   This command displays the path where the files would have been written, as well as the structure itself. Make sure the path in the first line of output matches your `integrations-extras` repository location.
```

{{< /code-block >}}

To speed up development, use the `-m/--marker` option to run integration tests only:

```
ddev test -m integration awesome
```

Your integration is almost complete. Next, add the necessary check assets.

## Create the check assets

> The set of assets created by the `ddev` scaffolding must be populated in order for a check to be considered in `integrations-extras`:

Reviewer: These assets were already created when the user ran ddev create in step 2, so I think the heading could say "Populate..." instead. 

Also I think "Populate integration tile assets" or something along those lines is clearer than "check"

{{< /code-block >}}

To speed up development, use the `-m/--marker` option to run integration tests only:
   ```
   ddev test -m integration awesome
   ```
Your integration is almost complete. Next, add the necessary check assets.

## Create the check assets

The set of assets created by the `ddev` scaffolding must be populated:

`spec.yaml`
: This is used to generate the `conf.yaml.example` using the `ddev` tooling (see the **Configuration template** tab below). For more information, see [Configuration specification][11].

`conf.yaml.example`
: This contains default (or example) configuration options for your Agent Check. **Do not edit this file by hand!** It is generated from the contents of `spec.yaml`. For more information, see the [Configuration file reference][12].

`metadata.csv`
: This contains the list of all metrics collected by your Agent Check. For more information, see the [Metrics metadata file reference][14].

`service_check.json`
: This contains the list of all Service Checks collected by your Agent Check. For more information, see the [Service check file reference][15].

For more information on the `README.md` and `manifest.json` files, see [Create a Tile][20].

> {{< tabs >}}

Author: ```suggestion
{{% /tab %}}
```

{{< /code-block >}}

To speed up development, use the `-m/--marker` option to run integration tests only:
   ```
   ddev test -m integration awesome
   ```
Your integration is almost complete. Next, add the necessary check assets.

## Create the check assets

The set of assets created by the `ddev` scaffolding must be populated:

`spec.yaml`
: This is used to generate the `conf.yaml.example` using the `ddev` tooling (see the **Configuration template** tab below). For more information, see [Configuration specification][11].

`conf.yaml.example`

> : This contains default (or example) configuration options for your Agent Check. **Do not edit this file by hand!** It is generated from the contents of `spec.yaml`. For more information, see the [Configuration file reference][12].

Reviewer 4: I would remove the exclamation mark.

`metadata.csv`
: This contains the list of all metrics collected by your Agent Check. For more information, see the [Metrics metadata file reference][14].

`service_check.json`
: This contains the list of all Service Checks collected by your Agent Check. For more information, see the [Service check file reference][15].

> For more information on the `README.md` and `manifest.json` files, see [Create a Tile][20].

Reviewer 2: @anawishnoff Also realized that we link them to the asset integration tile which also relinks to another page if they want more info. Should we just include them here too?

> For more information on the `README.md` and `manifest.json` files, see [Create a Tile][20].

Reviewer 2: ![Create an Agent Integration 2023-04-25 at 3 12 13 PM](https://user-images.githubusercontent.com/52013262/234392041-15043573-a364-4d5b-9642-0861076604ec.jpg)
![Create a Tile 2023-04-25 at 3 12 25 PM](https://user-images.githubusercontent.com/52013262/234392070-610c4a84-66b1-472b-a527-23118aa4811d.jpg)

> For more information on the `README.md` and `manifest.json` files, see [Create a Tile][20].

Reviewer: I find this confusing. The user already created everything they needed for a tile when they ran the `ddev create` command. 

These instructions seems to tell the user to fork/clone a GitHub repo and then **create an integration** with ddev, and then fork/clone a GitHub repo and **create a tile** with ddev, which would be redundant. The user will end up with two directories for two different integrations. 

Instead, can we clarify that the user just needs to populate the existing files they already created? 

For example, the "Create an API integration" page could have the (tile-only) ddev command , and the "Create a Tile" page could just walk through populating the scaffolding for both Agent and API integrations

{{< /code-block >}}

To speed up development, use the `-m/--marker` option to run integration tests only:
   ```
   ddev test -m integration awesome
   ```
Your integration is almost complete. Next, add the necessary check assets.

> ## Create the check assets

Reviewer: Re-adding my comment that was marked resolved:

These assets were already created when the user ran ddev create, so I think the heading could say "Populate..." instead.

Also I think "Populate integration tile assets" or something along those lines is clearer than "check"

Reviewer 3: addressed this by editing the last section of the agent integration doc to focus on populating the tile rather than "creating a tile", and added a warning in the create a tile doc that steps can be skipped if a file scaffolding was already created

```

{{% /tab %}}

{{% tab "Metadata" %}}

For this example, the Awesome integration doesn't provide any metrics, so in this case, the generated `awesome/metadata.csv` only contains only a row with the column names.

> {{% /tab %}}

Author: ```suggestion
{{% /tab %}}
{{< tabs >}}
```

ddev validate config --sync awesome
```

{{% /tab %}}
{{% tab "Metadata" %}}

> For this example, the Awesome integration doesn't provide any metrics, so in this case, the generated `awesome/metadata.csv` only contains only a row with the column names.

Reviewer: I find this tab confusing. It looks to be blank, and doesn't show the row with column names.

Reviewer: Just calling it "metadata" and not "metadata.csv" also makes it hard to follow. 

More than that, though, is there a reason why we are showing this "metadata" example next to "Configuration template"? There aren't examples for readme, the manifest, etc.

Reviewer 3: @hestonhoffman I think Alex's feedback above falls into your wheelhouse - can you provide context on why we only show previews for a few of the asset types? Perhaps we don't need to show previews at all?

  ```
</details>

## Create a tile and publish your integration

After you've created your Agent-based integration, populate the remainder of the required assets that will appear throughout your integration tile by following the instructions in [Create a Tile][20].

> Finally, open a pull request with your code on [integrations-extras] or [marketplace]. After you've created your pull request, automatic checks run to verify that your pull request is in good shape and contains all the required content to be updated.

Author: ```suggestion
Finally, open a pull request with your code on the [`integrations-extras`][21] or [`marketplace`][22] repository. After you've created your pull request, automatic checks run to verify that your pull request is in good shape and contains all the required content to be updated.
```

  ```
</details>

## Populate your tile and publish your integration

> Once you have created your Agent-based integration, populate the remaining [required assets][31] that  appear on your integration tile. For more information on populating these assets and best practices, see [Create a Tile][20].

Reviewer 5: ```suggestion
Once you have created your Agent-based integration, see the [Create a tile][20] documentation for information on populating the remaining [required assets][31] that appear on your integration tile, and opening a pull request.
```

> Finally, open a pull request with your code on the [`integrations-extras`][21] or [`marketplace`][22] repository. After you've created your pull request, automatic checks run to verify that your pull request is in good shape and contains all the required content to be updated.

Reviewer 5: ```suggestion
```

I don't think this paragraph is useful. I think creating a tile should be the last thing the user sees on this page - that page has much better instructions for opening the pull request - and the steps they have to perform before they open the PR - like the validation step.

[17]: https://docs.datadoghq.com/agent/
[18]: https://www.datadoghq.com/blog/programmatically-manage-your-datadog-integrations/
[19]: https://desktop.github.com/

> [20]: Create a tile 

Author: ```suggestion
[20]: https://docs.datadoghq.com/developers/integrations/create_a_tile
[21]: https://github.com/Datadog/integrations-extras
[22]: https://github.com/Datadog/marketplace
```
